src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+   * 
+   * @param placeId placeId to search openHours for
+   * @param index index to put the openHours in the HashMap
+   */
+  private void setOpenHours(int index, String placeId) throws IOException {
+    PlaceDetails placeDetails = TripServlet.getPlaceDetailsFromPlaceId(this.context, placeId);
+
+    if (placeDetails.openingHours == null) { 
+      this.openHours.put(index, null);    
+    } else if (placeDetails.openingHours.periods.length != 7) {
+      checkDayIsPresentInHours(placeDetails.openingHours, index);
+    } else {
+      this.openHours.put(index, placeDetails.openingHours.periods[this.intOfWeek]);
+    } 
+  }
+
+  /**
+   * If day in openingHours array is the same as the day searched for 
+   * check if open and close time are not null. If no hours found put null
+   * into openHours (represents open all day)
+   * 
+   * @param openingHours openHours for placeId on desired day
+   * @param index index to put result in OpenHours Map
+   */
+  private void checkDayIsPresentInHours(OpeningHours openingHours, int index) {
+      int adjustedWeekIndex;
+      // periods is numbered Sunday - Saturday, WeekdayText is Monday - Sunday
+      if (intOfWeek == 0) {
+        adjustedWeekIndex = 6;
+      } else {
+        adjustedWeekIndex = intOfWeek - 1;
+      }
+      // string representation of day searched
+      String desiredDayString = WEEKDAYS[adjustedWeekIndex];
+      for (int i = 0; i < openingHours.periods.length; i++) {
+        // get the day that the ith index in opening hours corresponds to
+        String day = openingHours.weekdayText[i].split(":")[0];
+
+        // check neither open nor closing time are null
+        boolean open = openingHours.periods[i].open != null;
+        boolean close = openingHours.periods[i].close != null;
+
+        if (day.equals(desiredDayString) && open && close) {
+          this.openHours.put(index, openingHours.periods[i]);
+          return;
+        } 
+      }
+    this.openHours.put(index, null);
+  }
+
+  /**
+   * Get travel time in seconds between two locations
+   * 
+   * @param origin start location
+   * @param destination end location
+   * @return time in seconds to travel from start to end
+   */
+  private int getTravelTimeMins(String origin, String destination) 
+      throws IOException {
+    // generate directions request w/ origin, destination by driving
+    DirectionsApiRequest directionsRequest = DirectionsApi.newRequest(this.context)
+        .originPlaceId(origin)
+        .destinationPlaceId(destination)
+        .mode(TravelMode.DRIVING);
+
+    // get directions Result from directionsRequest
+    DirectionsResult dirResult = TripServlet.getDirectionsResult(directionsRequest);
+
+    // get the time in minutes of the first the trip
+    DirectionsLeg leg  = dirResult.routes[TripServlet.ROUTE_INDEX].legs[0];
+    int travelTime = (int) leg.duration.inSeconds / TripServlet.SECONDS_IN_MIN;
+    return travelTime;
+  }
+
+  /**
+   * Getter functions
```

Comment:
nit: might be worth mentioning that all of these except for getFinalAnswer are for tests only.